pkg/backend/snapshot.go

```diff
@@ -49,96 +43,109 @@ type SnapshotPersister interface {
+	baseSnapshot *deploy.Snapshot // The base snapshot for this plan
 
-	completeOps      map[*resource.State]bool // The set of resources that have completed their operation
-	mutationRequests chan<- mutationRequest   // The queue of mutation requests, to be retired serially by the manager
-	cancel           chan bool                // A channel used to request cancellation of any new mutation requests.
-	done             <-chan error             // A channel that sends a single result when the manager has shut down.
-
-	refreshDeletes map[resource.URN]bool // The set of resources that have been deleted by a refresh in this plan.
+	// newResources is a map of resources that have been added to the snapshot in this plan, keyed by the resource
+	// state.  This is used to track the added resources and their operation IDs, allowing us too delete
+	// them later if necessary.
+	newResources       gsync.Map[*resource.State, uint64]
+	operationIDCounter atomic.Uint64 // A counter used to generate unique operation IDs for journal entries.
 }
 
 var _ engine.SnapshotManager = (*SnapshotManager)(nil)
 
-type mutationRequest struct {
-	mutator func() bool
-	result  chan<- error
-}
-
 func (sm *SnapshotManager) Close() error {
-	close(sm.cancel)
-	return <-sm.done
+	return sm.journal.Close()
 }
 
-// If you need to understand what's going on in this file, start here!
-//
-// mutate is the serialization point for reads and writes of the global snapshot state.
-// The given function will be, at the time of its invocation, the only function allowed to
-// mutate state within the SnapshotManager.
-//
-// Serialization is performed by pushing the mutator function onto a channel, where another
-// goroutine is polling the channel and executing the mutation functions as they come.
-// This function optionally verifies the integrity of the snapshot before and after mutation.
-//
-// The mutator may indicate that its corresponding checkpoint write may be safely elided by
-// returning `false`. As of this writing, we only elide writes after same steps with no
-// meaningful changes (see sameSnapshotMutation.mustWrite for details). Any elided writes
-// are flushed by the next non-elided write or the next call to Close.
-//
-// You should never observe or mutate the global snapshot without using this function unless
-// you have a very good justification.
-func (sm *SnapshotManager) mutate(mutator func() bool) error {
-	result := make(chan error)
-	select {
-	case sm.mutationRequests <- mutationRequest{mutator: mutator, result: result}:
-		return <-result
-	case <-sm.cancel:
-		return errors.New("snapshot manager closed")
+type JournalEntryKind int
+
+const (
+	JournalEntryBegin          JournalEntryKind = 0
+	JournalEntrySuccess        JournalEntryKind = 1
+	JournalEntryFailure        JournalEntryKind = 2
+	JournalEntryRefreshSuccess JournalEntryKind = 3
+	JournalEntryOutputs        JournalEntryKind = 4
+	JournalEntryRebase         JournalEntryKind = 5
```

Comment:
Should this be `JournalEntryWrite`? From the comments it seems that rebase was renamed to write at some point.